src/sinks/datadog/logs/sink.rs

```diff
@@ -137,6 +153,76 @@ pub fn normalize_event(event: &mut Event) {
     }
 }
 
+// Optionally for all other non-reserved fields, nest these under the `message` key. This is the
+// final step in having the event conform to the standard that the logs intake expects when an
+// event originates from an agent. Normalizing the events to the format prepared by the datadog
+// agent resolves any inconsistencies that would be observed when data flows through vector
+// before being ingested by the logs intake. This is because the logs intake interprets the
+// request with slight differences when this header and format are observed.
+pub fn normalize_as_agent_event(event: &mut Event) {
+    let log = event.as_mut_log();
+
+    // Extract the `message` object from the log if it already exists. Then JSON deserialize its
+    // value field. All non reserved fields placed at the root will be moved there.
+    let mut message = if let Some(outer_msg) = log.remove(event_path!(MESSAGE)) {
+        if let Value::Object(current_message) = outer_msg {
+            current_message // already JSON
+        } else {
+            // possible payload is stringified JSON
+            outer_msg
+                .as_bytes()
+                .and_then(|b| serde_json::from_slice::<ObjectMap>(b).ok())
+                .unwrap_or_else(|| {
+                    [(KeyString::from(MESSAGE), outer_msg)]
+                        .into_iter()
+                        .collect()
+                })
+        }
+    } else {
+        ObjectMap::default()
+    };
+
+    // Move all non reserved fields into a new object
+    if let Some(object_map) = log.as_map_mut() {
+        let mut collisions = ObjectMap::default();
+
+        let keys_to_move = object_map
+            .keys()
+            .filter(|key| {
+                DD_RESERVED_SEMANTIC_ATTRS
+                    .iter()
+                    .all(|(_, attr)| *attr != key.as_str())
```

Comment:
nit: Personally I'd use an `any` here, and possibly move the whole search into where `DD_RESERVED_SEMANTIC_ATTRS` is defined:
```suggestion
                !DD_RESERVED_SEMANTIC_ATTRS
                    .iter()
                    .any(|(_, attr)| *attr == key.as_str())
```